name: Re-run Flaky Workflows

on:
  workflow_run:
    workflows: [Backend, Driver Tests, E2E Tests, Frontend]
    types: [completed]
    branches: [master, 'release-**']

jobs:
  rerun-on-failure:
    name: 'Re-run ''${{ github.event.workflow_run.name }}'' workflow'
    runs-on: ubuntu-22.04
    if: ${{ github.event.workflow_run.conclusion == 'failure' }}
    steps:
      - name: Generate job summary
        run: |
          echo "# ${{ github.event.workflow_run.name }} workflow failed! :x:" >> $GITHUB_STEP_SUMMARY
          echo "" >> $GITHUB_STEP_SUMMARY
          echo "View the failed run attempt (#${{ github.event.workflow_run.run_attempt }}) using the following link:" >> $GITHUB_STEP_SUMMARY
          echo "${{ github.event.workflow_run.html_url }}" >> $GITHUB_STEP_SUMMARY
      - uses: actions/setup-node@v3
        with:
          node-version: '16'
      - run: npm install @slack/web-api
      - name: Trigger a re-run
        uses: actions/github-script@v6
        with:
          script: |
            const MAX_ATTEMPTS = 2;
            const ATTEMPT = ${{ github.event.workflow_run.run_attempt }};

            if (ATTEMPT <= MAX_ATTEMPTS) {
              github.rest.actions.reRunWorkflowFailedJobs({
                owner: context.repo.owner,
                repo: context.repo.repo,
                run_id: ${{ github.event.workflow_run.id }},
              });
            } else {
              // notify slack of repeated failure
              const { WebClient } = require('@slack/web-api');
              const slack = new WebClient('${{ secrets.SLACK_BOT_TOKEN }}');

              const author = `@${context.actor}`;
              const breaking_commit = context.sha;
              const branch = context.ref;

              await slack.chat.postMessage({
                channel: 'engineering-ci',
                blocks: [
                  {
                    "type": "header",
                    "text": {
                      "type": "plain_text",
                      "text": `:alert: CI is failing on ${branch} :alert:`,
                      "emoji": true,
                    }
                  },
                ],
                attachments: [{
                  color: '#f85149',
                  blocks: [
                    {
                      "type": "section",
                      "text": {
                        "type": "mrkdwn",
                        "text": `Commit <https://github.com/metabase/metabase/commit/${breaking_commit}|${breaking_commit.slice(0,7)}> by ${author} has failing tests on the \`${branch}\` branch . :sad-panda:\nPlease fix ASAP. Emoji this message when it's fixed.`
                      }
                    },
                    {
                      "type": "actions",
                      "elements": [
                        {
                          "type": "button",
                          "text": {
                            "type": "plain_text",
                            "text": "Failing Commit",
                            "emoji": true
                          },
                          "url": `https://github.com/metabase/metabase/commit/${breaking_commit}`
                        },
                        {
                          "type": "button",
                          "text": {
                            "type": "plain_text",
                            "text": "Recent commits to master",
                            "emoji": true
                          },
                          "url": `https://github.com/metabase/metabase/commits/${branch}`
                        }
                      ]
                    }
                  ]
                }]
              });
            }
